File: source/smc/generator/csharp/CSharpCodeGenerators/FSMClass.cs
namespace SMC.Generator.CSharp.CSharpCodeGenerators
{
    using System;
    using System.Linq;
    using System.Text;

    using SMC.FsmRep;
    using SMC.Generator.CSharp;

    public class FSMClass : CSharpCodeGenerator
    {
        #region Public Methods

        public override string GenerateCode(SMCSharpGenerator gen)
        {
            var stateMap = gen.StateMap;
            var buff = new StringBuilder();

            AddClassHeader(buff);
            BeginClassDeclaration(buff, stateMap);
            AddFields(buff, stateMap);
            AddClassMembers(gen, buff);

            AddStateClasses(gen, buff);

            CloseClassDeclaration(buff);

            return buff.ToString();
        }

        #endregion

        #region Methods

        private static void AddClassHeader(StringBuilder buff)
        {
            buff.AppendLine("/// <summary>")
                .AppendLine("/// This is the Finite State Machine class")
                .AppendLine("/// <summary>");
        }

        private static void BeginClassDeclaration(StringBuilder buff, StateMap stateMap)
        {
            var className = stateMap.Name;
            var superClassName = stateMap.ContextName;
            var qualifier = IsAbstract(stateMap) ? "abstract " : string.Empty;

            buff.AppendLine($"public {qualifier}class {className} : {superClassName}")
                    .AppendLine("{");
        }

        private static bool IsAbstract(StateMap stateMap) => stateMap.Pragma.Any(p => Matches(p.Trim(), "abstract"));

        private static bool Matches(string a, string b) => string.Compare(a, b, StringComparison.OrdinalIgnoreCase) == 0;

        private static void AddFields(StringBuilder buff, StateMap stateMap)
        {
            buff.AppendLine("    #region Fields")
                .AppendLine()
                .AppendLine($"    private static string version = \"{stateMap.Version}\";")
                .AppendLine()
                .AppendLine("    private State currentState;")
                .AppendLine()
                .AppendLine("    #endregion")
                .AppendLine();
        }

        private static void AddClassMembers(SMCSharpGenerator gen, StringBuilder buff) => AddFromGenerators(gen, CSharpCodeGeneratorBuilder.Instance.StateMachineGenerators, buff);

        private static void CloseClassDeclaration(StringBuilder buff) => buff.AppendLine("}");

        private static void AddStateClasses(SMCSharpGenerator gen, StringBuilder buff) => AddFromGenerators(gen, CSharpCodeGeneratorBuilder.Instance.StateGenerators, buff);

        #endregion
    }
}

File: source/smc/generator/csharp/CSharpCodeGenerators/FSMClass.java
package smc.generator.csharp.CSharpCodeGenerators;

import smc.generator.csharp.SMCSharpGenerator;
import smc.generator.csharp.CSharpCodeGeneratorBuilder;

import java.util.List;

public class FSMClass extends CSharpCodeGenerator
{
    public String generateCode(SMCSharpGenerator gen)
    {
        StringBuffer buff = new StringBuffer();
        String fsmName = gen.getStateMap().getName();
        buff.append(printSeparator(0));
        buff.append("/// <summary>\n");
        buff.append("/// This is the Finite State Machine class\n");
        buff.append("/// <summary>\n");
        buff.append("public class " + fsmName + " : " + gen.getStateMap().getContextName() + "\n");
        buff.append("{\n");

        buff.append("    #region Fields\n");
        buff.append("\n");
        buff.append("    private State currentState;\n");
        buff.append("    private static string version = \"" + gen.getStateMap().getVersion() + "\";\n");
        buff.append("\n");
        buff.append("    #endregion\n");
        buff.append("\n");

        try
        {
            List generators = CSharpCodeGeneratorBuilder.cSharpCode.cSharpFSMInstances();
            for(int i = 0; i !=generators.size(); i++)
            {
				System.out.println(generators.get(i).getClass());
                CSharpCodeGenerator code = (CSharpCodeGenerator)generators.get(i);
                buff.append(code.generateCode(gen));
            }
        }
        catch(Exception e )
        {}

        buff.append("}\n");
        buff.append("\n");
        
        try
        {
            List generators = CSharpCodeGeneratorBuilder.cSharpCode.cSharpFSMClassesInstances();
            for(int i = 0 ; i !=generators.size();i++)
            {
                CSharpCodeGenerator code = (CSharpCodeGenerator)generators.get(i);
                buff.append(code.generateCode(gen));
            }
        }
        catch(Exception e )
        {}

        return buff.toString();
    }
}
